### PR TITLE
Small initial pull request

### PR DESCRIPTION
**Pull request by mattagape**
This is an initial PR for @skeating to look at, and check I am on the right path! We have agreed to use my `mattdoc` branch to do PRs little and often.
I have marked a number of queries with TODO.

### CODE REVIEW

File: generator/parseXML/ParseXML.py [hunk 1]
```diff
@@ -79,6 +79,7 @@ def to_bool(v):
         if v is None:
             return False
         return v.lower() in ("yes", "true", "1")
+        # TODO what is returned if v.lower() is "false"?
```
Comment (skeating) [inline generator/parseXML/ParseXML.py]: Good question. I suggest that where I don't answer a question you create a github issue for the 'question'. You could then take out the TODO.

File: generator/parseXML/ParseXML.py [hunk 2]
```diff
@@ -184,24 +185,38 @@ def get_enum_name(self, node):
...
+        Used to get information from a <mapping> node
+        e.g.
+        <mapping name="DimensionDescription" package="numl"/>
+
+        :param node: the <mapping> node to interrogate
+        :return: returns [name, package] tuple
+        '''
         name = ''
-        package = ''
+        package = ''  # TODO Maybe better as None?
```
Comment (skeating) [inline generator/parseXML/ParseXML.py:302]: Agreed

File: generator/parseXML/ParseXML.py [hunk 3]
```diff
@@ -184,24 +185,38 @@ def get_enum_name(self, node):
...
 
 
     @staticmethod
     def get_loclass_name_value(self, node, name):
         xml_loclass_name = ''
         temp = self.get_value(node, name)
-        # we expect this to be camel case starting with lower
+        # we expect this to be camelCase starting with lower
         if temp is not None:
             xml_loclass_name = strFunctions.upper_first(temp)
```
Comment (skeating) [inline generator/parseXML/ParseXML.py]: Can you add an issue for this class as it says camelCase with lower first but code does upper_first

File: generator/parseXML/ParseXML.py [hunk 4]
```diff
@@ -247,8 +262,17 @@ def get_typecode(self, node):
         return typecode
 
     def analyse_enumname(self, enum_name):
+        """
+        Pulls out "parts" of the name of an <enum> node, depending on
+        whether an upper-case letter is found in the name or not.
+
+        TODO: why?
```
Comment (skeating) [inline generator/parseXML/ParseXML.py:382]: SK needs to look at this and came back and answer

File: generator/parseXML/ParseXML.py [hunk 5]
```diff
@@ -295,6 +319,7 @@ def find_lo_element(elements, name):
     # Functions for standardizing types/name etc
 
     def standardize_types(self, attrib_type):
+        # TODO A dictionary would work well here
```
Comment (skeating) [inline generator/parseXML/ParseXML.py]: Agreed.

File: generator/parseXML/ParseXML.py [hunk 6]
```diff
@@ -651,6 +689,22 @@ def get_dependency_information(self, node):
...
+
+        Example of a <language> node starting element:
+        <language name="SBGN" baseClass="SbgnBase" documentClass="SbgnDocument"
+         prefix="Sbgn" libraryName="libSBGN"
+         annotationElementName="Extension" topLevelElementName="Extension"
+         isPackage="false" uses_ASTNode="true" uses_XMLNode="true">
+
+        The <language> node can also contain nodes of at least the following
+        types: <library_version>, <language_version> and <dependencies>.
+        These are also interrogated in this function.
```
Comment (skeating) [inline generator/parseXML/ParseXML.py:843]: These would be better as functions for interrogating each type of node.

File: generator/parseXML/ParseXML.py [hunk 7]
```diff
@@ -779,29 +851,36 @@ def parse_deviser_xml(self):
         global_variables.add_additional_declaration(add_declarations)
 
         # get package information
+        # TODO Should these numbers be hardcoded here? Maybe better in their own file?
```
Comment (skeating) [inline generator/parseXML/ParseXML.py]: Add an issue for this so it gets discussed.

File: generator/parseXML/ParseXML.py (hunk 1 above)
Comment (mattagape) [reply generator/parseXML/ParseXML.py]: Now issue 15

File: generator/parseXML/ParseXML.py (hunk 2 above)
Comment (mattagape) [reply generator/parseXML/ParseXML.py:302]: Do we need to change the calling code though? (In function `parse_deviser_xml()`)

File: generator/parseXML/ParseXML.py (hunk 3 above)
Comment (mattagape) [reply generator/parseXML/ParseXML.py]: Now issue 16

File: generator/parseXML/ParseXML.py (hunk 7 above)
Comment (mattagape) [reply generator/parseXML/ParseXML.py]: Now issue 17

File: generator/parseXML/ParseXML.py (hunk 5 above)
Comment (mattagape) [reply generator/parseXML/ParseXML.py]: OK I have changed it

File: generator/parseXML/ParseXML.py (hunk 6 above)
Comment (mattagape) [reply generator/parseXML/ParseXML.py:843]: Do you mean we need to add these as new functions?

Comment (skeating) [reply generator/parseXML/ParseXML.py:843]: Yes that would be cleaner

File: generator/parseXML/ParseXML.py (hunk 2 above)
Comment (skeating) [reply generator/parseXML/ParseXML.py:302]: Shouldnt have thought so !

File: generator/parseXML/ParseXML.py (hunk 4 above)
Comment (skeating) [reply generator/parseXML/ParseXML.py:382]: So this one is due to the different styles of naming in an enum. We want the code to end up with a name that is
FOO_BAR_XYZ but we cannot guarantee that these have been entered this way. I used cname (classname) and tname (typename) to distinguish.

So you might have BOUNDARYKIND or BOUNDARY_KIND or BoundaryKind; which are all the classname and so we want

typedef enum
{
  SPATIAL_BOUNDARYKIND_ROBIN_VALUE_COEFFICIENT                        /*!< The spatial boundarykind is @c "Robin_valueCoefficient". */
, SPATIAL_BOUNDARYKIND_ROBIN_INWARD_NORMAL_GRADIENT_COEFFICIENT       /*!< The spatial boundarykind is @c "Robin_inwardNormalGradientCoefficient". */
, SPATIAL_BOUNDARYKIND_ROBIN_SUM                                      /*!< The spatial boundarykind is @c "Robin_sum". */
, SPATIAL_BOUNDARYKIND_NEUMANN                                        /*!< The spatial boundarykind is @c "Neumann". */
, SPATIAL_BOUNDARYKIND_DIRICHLET                                      /*!< The spatial boundarykind is @c "Dirichlet". */
, SPATIAL_BOUNDARYKIND_INVALID                                        /*!< Invalid BoundaryKind value. */
} BoundaryKind_t;

so first two parts are always the same and then the part after second _ is value - which might have underscores.

Hope this makes sense.

**Review by skeating — APPROVED** [review]
have left some answers but nothing to do

File: generator/parseXML/ParseXML.py (hunk 4 above)
Comment (mattagape) [reply generator/parseXML/ParseXML.py:382]: Thanks Sarah. I will add this comment into the doc for that function

File: generator/parseXML/ParseXML.py (hunk 6 above)
Comment (mattagape) [reply generator/parseXML/ParseXML.py:843]: I will add a TODO comment to the code noting this, for later action.

**Pull request merged**